### PR TITLE
[Feat] SenseNova-U1 online dynamic batching

### PR DESCRIPTION
**Pull request by 86MaxCao**
## Purpose

Implements the **online dynamic batching** item from the [SenseNova-U1 Optimization Roadmap](https://github.com/vllm-project/vllm-omni/issues/3514). Instead of running the full denoising loop inside `forward()`, the scheduler drives the loop one timestep at a time via the `SupportsStepExecution` protocol, allowing requests with different prompts, CFG scales, and step counts to share the same GPU batch.

**Main Work**:
- Phase 1 — Step-wise execution interface 
Implemented SupportsStepExecution protocol on SenseNovaU1Pipeline (prepare_encode, denoise_step, step_scheduler, post_decode). AR prefill (T2I and IT2I) moved into prepare_encode; denoising loop is now driven externally by the framework's StepScheduler.

- Phase 2 — Batched forward
Replaced forward_gen_varlen with forward_gen_batched at all four levels (SenseNovaU1Attention, SenseNovaU1DecoderLayer, SenseNovaU1Model, SenseNovaU1ForCausalLM). New path uses padded [B, max_S, D] tensors + 2D boolean attention mask through _run_attn_bshd, removing the direct flash_attn import and keeping the code backend-agnostic.

- Phase 3 — Per-request CFG
_batched_denoise_step handles mixed CFG scales per request — each request applies its own cfg_scale / img_cfg_scale / cfg_norm independently. Supports T2I (two-way CFG) and IT2I (dual CFG with img_cond) requests in the same batch. Bypasses InputBatch._prepare_cfg_scalars homogeneity check by setting do_true_cfg=False and managing CFG entirely inside the pipeline.

**Test Environment**
Model: **SenseNova-U1-8B-MoT**
Hardware: **NVIDIA H20 (96 GiB)**

## Test Plan 1 — Unit & E2E tests

```bash
# Scheduler unit tests (CPU, no model weights needed)
pytest tests/diffusion/test_diffusion_scheduler.py -v                          # 51 tests

# E2E step execution tests (requires GPU + model weights)
pytest tests/e2e/offline_inference/test_sensenova_u1_text2img_expansion.py -k step_execution -v
pytest tests/e2e/offline_inference/test_sensenova_u1_img2img_expansion.py -k step_execution -v
```

## Test Result 1

All tests passed.

| # | Test | Type | Tests | Status |
|---|------|------|-------|--------|
| 1 | `test_diffusion_scheduler.py` | Unit (CPU) | 51 | PASS |
| 2 | `test_sensenova_u1_text2img_step_execution` | E2E GPU | 1 | PASS |
| 3 | `test_sensenova_u1_img2img_step_execution` | E2E GPU | 1 | PASS |

## Test Plan 2 — cache_dit vs step-execution (bit-exact verification)

Step execution mode is enabled via `step_execution=True` in `OmniDiffusionConfig`. Prior to merging upstream, bit-exact equivalence between cache_dit (request mode) and step-execution was verified using `end2end.py` (since removed in upstream PR #5201). After merge, e2e tests with the real model confirm step execution produces correct results.

## Test Result 2

cache_dit vs Step-execution: **bit-exact** (MSE=0.0000)

| cache_dit | step-execution |
|---|---|
|  |  |

## Known Limitations

- **Step-execution does not yet support Cache-DiT.** The current Cache-DiT architecture uses pipeline-level global state, which is incompatible with per-request batching. This may be resolved once #3473 (DiT Cache Pool) lands.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [x] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by princepride** [comment]
Woo, so many lines change😂, I think we don't need so many unit tests, we can maintain only one smoking test ensure this feature still working.

**Comment by 86MaxCao** [comment]
> Woo, so many lines change😂, I think we don't need so many unit tests, we can maintain only one smoking test ensure this feature still working.


Thanks for the feedback! I've removed all the verbose unit/integration tests under tests/diffusion/models/sensenova_u1/
 (~3200 lines deleted) and kept only the two e2e smoke tests:

- tests/e2e/offline_inference/test_sensenova_u1_text2img.py
- tests/e2e/offline_inference/test_sensenova_u1_img2img.py

I also kept tests/diffusion/test_diffusion_scheduler.py (193 lines, CPU-only) since it covers the generic scheduler framework logic shared across models — let me know if you'd prefer I remove that one too.

**Comment by princepride** [comment]
@alex-jw-brooks @lishunyang12 can you help me test it?

File: vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py [hunk 1]
```diff
@@ -1476,6 +1486,572 @@ def _run_denoising_loop(self, ns, caches, p, think_text="") -> DiffusionOutput:
...
+            cfg_interval=tuple(extra_args.get("cfg_interval", (0.0, 1.0))),
+            batch_size=1,
+            seed=seed,
+            think_mode=bool(extra_args.get("think", False)),
+            t_eps=float(extra_args.get("t_eps", 0.02)),
+        )
+
+    def _build_t2i_caches(self, p: SimpleNamespace, ns: SimpleNamespace) -> dict[str, Any]:
+        """Build KV caches for text-to-image generation."""
+        think_content = "<think>\n" if p.think_mode else "<think>\n\n</think>\n\n" + IMG_START_TOKEN
```
Comment (alex-jw-brooks) [inline vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py]: Can you pull this out to a var, similar to the image tokens? We should avoid hardcoding here, in `think_off`, etc

File: vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py [hunk 2]
```diff
@@ -1476,6 +1486,572 @@ def _run_denoising_loop(self, ns, caches, p, think_text="") -> DiffusionOutput:
...
+
+        needs_cfg = not (p.cfg_scale == 1 and p.img_cfg_scale == 1)
+        needs_img_cond = needs_cfg and (p.img_cfg_scale == 1 or p.cfg_scale != p.img_cfg_scale)
+        needs_uncond = needs_cfg and p.img_cfg_scale != 1
+
+        query_cond = self._build_it2i_query(p.prompt, images_info, p.think_mode)
+        embeds_cond, idx_cond, mask_cond = self._build_it2i_inputs(query_cond, pixel_values, grid_hw)
+
+        if needs_img_cond:
+            img_only_prompt = "<image>" * grid_hw.shape[0]
```
Comment (alex-jw-brooks) [inline vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py]: I don't think `<image>` should be used as a placeholder here, since it's a common placeholder for other visual language models etc, but not for this model. So if this is broken in a future refactor, it will be confusing.

Can you refactor this to just append tokens in a list "".join() the results instead of using `<image>` + .replace?

File: vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py [hunk 3]
```diff
@@ -1476,6 +1486,572 @@ def _run_denoising_loop(self, ns, caches, p, think_text="") -> DiffusionOutput:
...
+            caches = self._build_t2i_caches(p, ns)
+
+        # Populate state fields consumed by InputBatch / step_scheduler.
+        # timesteps[:-1] gives exactly num_steps entries so that
+        # DiffusionRequestState.total_steps == num_steps.
+        state.latents = ns.image_prediction
+        state.timesteps = ns.timesteps[:-1]
+        state.step_index = 0
+        # U1 handles CFG internally via state.extra["p"].cfg_scale,
+        # not via InputBatch's do_true_cfg mechanism. Setting False
```
Comment (alex-jw-brooks) [inline vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py:1706]: I'm not very familiar with this detail - do you know why there isn't a unified way of handling this?

File: vllm_omni/diffusion/models/sensenova_u1/sensenova_u1_transformer.py [hunk 4]
```diff
@@ -491,6 +491,66 @@ def forward_gen(self, hidden_states, indexes, attention_mask, past_key_values=No
...
+        hidden_states: torch.Tensor,
+        indexes: torch.Tensor,
+        prefix_kv_list: list[tuple[torch.Tensor, torch.Tensor]],
+        cu_seqlens_q: torch.Tensor,
+        cu_seqlens_k: torch.Tensor,
+        max_seqlen_q: int,
+        max_seqlen_k: int,
+    ) -> torch.Tensor:
+        """Varlen generation path for batched requests with different prefix lengths."""
+        from flash_attn import flash_attn_varlen_func
```
Comment (alex-jw-brooks) [inline vllm_omni/diffusion/models/sensenova_u1/sensenova_u1_transformer.py]: We should not add a separate `forward_len` path for batched inputs, especially since this implementation is dependent on Flash Attention. The FA backend may also import the flash attention func / varlen func from a variety of places, so this may break in unexpected ways.

The Diffusion backend should be able to handle varlen inputs if you provide an attention mask (you can see the relevant FA code [here](https://github.com/86MaxCao/vllm-omni/blob/b130b84f8df0864187ed8693a6675691a3eacbc3/vllm_omni/diffusion/attention/backends/flash_attn.py#L203). Can you refactor this to build the mask so that the diffusion backend handles this instead of making a separate pass for batch fwd?

File: vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py [hunk 5]
```diff
@@ -1476,6 +1486,572 @@ def _run_denoising_loop(self, ns, caches, p, think_text="") -> DiffusionOutput:
...
+        img = images[0] if images else None
+
+        think_text = state.extra.get("think_text", "")
+        custom: dict[str, Any] = {}
+        if think_text:
+            custom["think_text"] = think_text
+
+        # Cleanup KV caches and step state
+        caches = state.extra.get("caches", {})
+        for key in ("cond", "uncond", "img_cond"):
```
Comment (alex-jw-brooks) [inline vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py:2024]: There are a lot of checks against raw strings throughout this file that are hardcoded in line, can you refactor to clean these up? i.e., use more structured dataclasses or at least vars so that we don't accidentally miss an occurence of a string if this code changes in the future

File: vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py [hunk 6]
```diff
@@ -1476,6 +1486,572 @@ def _run_denoising_loop(self, ns, caches, p, think_text="") -> DiffusionOutput:
...
+            else:
+                x_pred = self.fm_modules["fm_head"](h_i[:, -image_token_num:].view(B, L, -1)).view(B, L, -1)
+
+            v_pred = (x_pred - z) / (1 - t).clamp_min(self.top_cfg.t_eps)
+            results.append(v_pred)
+            offset += q_lens[i]
+
+        return results
+
+    def _batched_denoise_step(
```
Comment (alex-jw-brooks) [inline vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py:1878]: I think this should try to share more code with the non-batched path since non-batched is ultimately just the degenerate case

File: vllm_omni/diffusion/data.py [hunk 7]
```diff
@@ -619,6 +619,11 @@ class OmniDiffusionConfig:
     # Supplementary model specific parameters
     extras: dict[str, Any] = Field(default_factory=dict)
 
+    # Fields in SamplingParamsKey to ignore when batching requests.
+    # Pipelines that handle heterogeneous batches (e.g. different resolutions)
+    # can set this to ["height", "width"] so the scheduler allows mixing.
+    heterogeneous_batch_fields: list[str] = field(default_factory=list)
```
Comment (alex-jw-brooks) [inline vllm_omni/diffusion/data.py]: I am opposed to adding this. It offers another place for pipelines to diverge from each other, and there are already a lot of places things can vary in pipelines between the normal pipeline approach, the stepwise approach, and now models like cosmo3 where one pipeline is meant to support multiple tasks. I think models implementing stepwise need to do so in such a way that the `SamplingParamsKey` is consistent. Then, changes to it to enable better batching should be made across all pipelines so that behavior is uniform for all models.

@lishunyang12 @princepride any thoughts on this?

**Review by alex-jw-brooks — CHANGES_REQUESTED** [review]

File: vllm_omni/diffusion/data.py (hunk 7 above)
Comment (princepride) [reply vllm_omni/diffusion/data.py]: I agree with you. @86MaxCao what do you think?

File: vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py (hunk 1 above)
Comment (86MaxCao) [reply vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py]: Done. Defined module-level constants `THINK_ON` and `THINK_OFF` (lines 70–71) and replaced all inline occurrences throughout the file.

File: vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py (hunk 2 above)
Comment (86MaxCao) [reply vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py]: Done. Refactored the IT2I `img_cond` blocks (both the non-batched and batched paths) to build the token sequence directly with a list + `"".join()`, eliminating the `<image>` placeholder entirely.

File: vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py (hunk 3 above)
Comment (86MaxCao) [reply vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py:1706]: InputBatch stores CFG as batch-level scalars (do_true_cfg: bool, true_cfg_scale: float), but U1 needs per-request heterogeneous CFG: T2I uses two-way CFG with cfg_zero_star, IT2I uses three-way dual CFG (cfg_scale + img_cfg_scale), and both scales are caller-specified. So I set do_true_cfg=False and handle CFG per-request in _batched_denoise_step.

Happy to file an issue to extend InputBatch with per-request CFG vectors + pluggable combination functions.

File: vllm_omni/diffusion/models/sensenova_u1/sensenova_u1_transformer.py (hunk 4 above)
Comment (86MaxCao) [reply vllm_omni/diffusion/models/sensenova_u1/sensenova_u1_transformer.py]: Done. Replaced forward_gen_varlen with forward_gen_batched at all four levels. New path uses padded [B, max_S, D] tensors + 2D boolean mask through _run_attn_bshd, removing the direct flash_attn import and keeping the code backend-agnostic.

File: vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py (hunk 5 above)
Comment (86MaxCao) [reply vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py:2024]: Done. Extracted 9 module-level constants for KV cache keys (lines 73–81): `COND`, `UNCOND`, `IMG_COND`, `IDX_COND`, `IDX_UNCOND`, `IDX_IMG_COND`, `MASK_COND`, `MASK_UNCOND`, `MASK_IMG_COND`. All inline string occurrences (~60 places) now reference these constants.

File: vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py (hunk 6 above)
Comment (86MaxCao) [reply vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py:1878]: Done. Both paths now share the same _combine_cfg logic. The non-batched path calls it through _denoise_step → _denoise_step_t2i / _denoise_step_it2i; the batched path calls it directly after the batched forward pass.

File: vllm_omni/diffusion/data.py (hunk 7 above)
Comment (86MaxCao) [reply vllm_omni/diffusion/data.py]: Agreed — removed `heterogeneous_batch_fields` from `OmniDiffusionConfig` and the related `_keys_match_ignoring` helper from the scheduler. Same-resolution batching only for now.

Regarding the unified approach you mentioned — where would you suggest adding support for heterogeneous-resolution batching across all pipelines? I'm happy to file an issue and contribute to the follow-up implementation.

That said, if this isn't a priority in the near term, I could also handle it locally by overriding `_make_request_state` in a model-specific scheduler subclass to normalize resolution fields in the `SamplingParamsKey` — this wouldn't touch any shared code. Either way works for me — let me know which direction you'd prefer.

**Comment by 86MaxCao** [comment]
@alex-jw-brooks 

Sorry for the late reply — just got off work.

Thanks for the detailed review!

I've addressed all comments and force-pushed the updated version.

**Comment by princepride** [comment]
@86MaxCao Can you help resolve the conflicts?

**Comment by 86MaxCao** [comment]
@princepride @alex-jw-brooks 

Sorry for the delayed response. I've resolved the merge conflicts with the latest upstream/main. All tests pass, including scheduler unit tests and e2e step execution tests with the real model. Please let me know if there's anything else that needs to be addressed.
